src/main/java/org/dice_research/opal/launuts/lau/ExcelParser.java

```diff
@@ -0,0 +1,190 @@
+
+import org.apache.commons.collections4.ListUtils;
+import org.apache.jena.ext.com.google.common.collect.Iterables;
+import org.apache.poi.ss.usermodel.*;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.*;
+
+/**
+ * Generic Parser to parse all the sheet from the xlsx.
+ * Used instead of {@link LauCsvParser}.
+ *
+ * @author Vikrant Singh
+ */
+
+
+public class ExcelParser implements LauReaderInterface {
+
+    public static final String FILE_NAME = "EU-28-LAU-2019-NUTS-2016.xlsx";
+
+    List<LauContainer> lauContainerList = new LinkedList<LauContainer>();
+    private HashMap<String, HashMap<String, LauContainer>> index;
+
+    private Workbook workbook;
+    private Sheet sheet;
+    private Map<String, List<String>> getCodes = new HashMap<>();
+    private HashMap<String, Integer> getkeys = new HashMap<String, Integer>();
+    private HashMap<String, LauContainer> laucodeToLauContainerMap = new HashMap<String, LauContainer>();
+    private List<String> countryIds;
+    private boolean parsed = false;
+
+    @Override
+    public LauReaderInterface setLauSourceDirectory() throws IOException {
+
+        workbook = WorkbookFactory.create(new File(FILE_NAME));
+        return this;
+    }
+
+    @Override
+    public List<String> getCountryIds() throws LauReaderException {
+
+        if (!parsed)
+            parse();
+
+        return countryIds;
+    }
+
+    @Override
+    public Map<String, List<String>> getCodes(String countryId) throws LauReaderException {
+
+        if (getCountryIds().contains(countryId)) {
+            sheet = workbook.getSheet(countryId);
+
+            if (!parsed) {
+                parse();
+            }
+        }
+
+        return getCodes;
+
+    }
+
+    @Override
+    public LauContainer getData(String nutsCode, String lauCode) throws LauReaderException {
+        LauContainer container = null;
+
+        if (!parsed) {
+            parse();
+        }
+
+        Map<String, LauContainer> lauCodeToContainerMap;
+        if (index.containsKey(nutsCode)) {
+            lauCodeToContainerMap = index.get(nutsCode);
+            container = lauCodeToContainerMap.get(lauCode);
+        }
+
+        return container;
+    }
+
+    @Override
+    public HashMap<String, Integer> getKeys() throws LauReaderException {
+
+        if (!parsed) {
+            parse();
+        }
+
+        return getkeys;
+    }
+
+    private void parse() throws LauReaderException {
+
+        String countryId = "";
+        index = new HashMap<String, HashMap<String, LauContainer>>();
+        List<String> oldLauCodes = new ArrayList<>();
+        String nutsCode = null;
+        List<String> lauCodes = new ArrayList<>();
+
+        if (countryIds == null) {
+            countryIds = new LinkedList<String>();
+            for (int i = 0; i < workbook.getNumberOfSheets(); i++) {
+                countryId = workbook.getSheetAt(i).getSheetName();
+                if (countryId.length() == 2)
```

Comment:
Why are we checking length? Is this for validation? I guess it would be better if we use regex for validation.

Reply:
Because we are using the country's abbreviations till 2 values, it's a simple way, and a particular regex cannot be set, as there are some sheets, which are not countries and will pass the regex, but does not have length 2. So I kept it plain and simple which works every time.

Reply:
Regex checks for length as well.  This regex is plain and simple -->> **^\D{2}$**
Checks for two non-digit characters of length 2.

Reply:
Though your idea is good, I feel, its an overhead to use pattern and matcher class for a small check.